commit 9319235a5506d6261abcded5680dce641b4df8d5
Author: CharlyCst <charly.castes@epfl.ch>
Date:   Sun Jul 27 19:26:01 2025 +0200

    Add rv8 benchmarks

diff --git a/examples/CMakeLists.txt b/examples/CMakeLists.txt
index 6806081..9bbe710 100644
--- a/examples/CMakeLists.txt
+++ b/examples/CMakeLists.txt
@@ -32,6 +32,10 @@ add_custom_target("examples")
 
 # add all examples below
 add_subdirectory(hello)
+add_subdirectory(bigint)
+add_subdirectory(dhrystone)
+add_subdirectory(miniz)
+add_subdirectory(primes)
+add_subdirectory(sha512)
 add_subdirectory(hello-native)
-add_subdirectory(attestation)
 add_subdirectory(tests)
diff --git a/examples/aes/CMakeLists.txt b/examples/aes/CMakeLists.txt
new file mode 100644
index 0000000..3617325
--- /dev/null
+++ b/examples/aes/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin aes)
+set(eapp_bin_path "eapp/aes")
+set(host_bin aes-runner)
+set(host_src host/host.cpp)
+set(package_name "aes.ke")
+set(package_script "./aes-runner aes eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/aes/eapp/.placeholder b/examples/aes/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/aes/host/host.cpp b/examples/aes/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/aes/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/bigint/CMakeLists.txt b/examples/bigint/CMakeLists.txt
new file mode 100644
index 0000000..1c4828e
--- /dev/null
+++ b/examples/bigint/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin bigint)
+set(eapp_bin_path "eapp/bigint")
+set(host_bin bigint-runner)
+set(host_src host/host.cpp)
+set(package_name "bigint.ke")
+set(package_script "./bigint-runner bigint eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/bigint/eapp/.placeholder b/examples/bigint/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/bigint/host/host.cpp b/examples/bigint/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/bigint/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/dhrystone/CMakeLists.txt b/examples/dhrystone/CMakeLists.txt
new file mode 100644
index 0000000..241b0f0
--- /dev/null
+++ b/examples/dhrystone/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin dhrystone)
+set(eapp_bin_path "eapp/dhrystone")
+set(host_bin dhrystone-runner)
+set(host_src host/host.cpp)
+set(package_name "dhrystone.ke")
+set(package_script "./dhrystone-runner dhrystone eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/dhrystone/eapp/.placeholder b/examples/dhrystone/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/dhrystone/host/host.cpp b/examples/dhrystone/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/dhrystone/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/iozone/CMakeLists.txt b/examples/iozone/CMakeLists.txt
new file mode 100644
index 0000000..0e1b20a
--- /dev/null
+++ b/examples/iozone/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin iozone)
+set(eapp_bin_path "eapp/iozone")
+set(host_bin iozone-runner)
+set(host_src host/host.cpp)
+set(package_name "iozone.ke")
+set(package_script "./iozone-runner iozone eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/iozone/eapp/.placeholder b/examples/iozone/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/iozone/host/host.cpp b/examples/iozone/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/iozone/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/miniz/CMakeLists.txt b/examples/miniz/CMakeLists.txt
new file mode 100644
index 0000000..c81cae2
--- /dev/null
+++ b/examples/miniz/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin miniz)
+set(eapp_bin_path "eapp/miniz")
+set(host_bin miniz-runner)
+set(host_src host/host.cpp)
+set(package_name "miniz.ke")
+set(package_script "./miniz-runner miniz eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/miniz/eapp/.placeholder b/examples/miniz/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/miniz/host/host.cpp b/examples/miniz/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/miniz/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/norx/CMakeLists.txt b/examples/norx/CMakeLists.txt
new file mode 100644
index 0000000..422b8a4
--- /dev/null
+++ b/examples/norx/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin norx)
+set(eapp_bin_path "eapp/norx")
+set(host_bin norx-runner)
+set(host_src host/host.cpp)
+set(package_name "norx.ke")
+set(package_script "./norx-runner norx eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/norx/eapp/.placeholder b/examples/norx/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/norx/host/host.cpp b/examples/norx/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/norx/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/primes/CMakeLists.txt b/examples/primes/CMakeLists.txt
new file mode 100644
index 0000000..e8af598
--- /dev/null
+++ b/examples/primes/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin primes)
+set(eapp_bin_path "eapp/primes")
+set(host_bin primes-runner)
+set(host_src host/host.cpp)
+set(package_name "primes.ke")
+set(package_script "./primes-runner primes eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/primes/eapp/.placeholder b/examples/primes/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/primes/host/host.cpp b/examples/primes/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/primes/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/qsort/CMakeLists.txt b/examples/qsort/CMakeLists.txt
new file mode 100644
index 0000000..9035f56
--- /dev/null
+++ b/examples/qsort/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin qsort)
+set(eapp_bin_path "eapp/qsort")
+set(host_bin qsort-runner)
+set(host_src host/host.cpp)
+set(package_name "qsort.ke")
+set(package_script "./qsort-runner qsort eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/qsort/eapp/.placeholder b/examples/qsort/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/qsort/host/host.cpp b/examples/qsort/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/qsort/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/sha512/CMakeLists.txt b/examples/sha512/CMakeLists.txt
new file mode 100644
index 0000000..7cad2ee
--- /dev/null
+++ b/examples/sha512/CMakeLists.txt
@@ -0,0 +1,35 @@
+set(eapp_bin sha512)
+set(eapp_bin_path "eapp/sha512")
+set(host_bin sha512-runner)
+set(host_src host/host.cpp)
+set(package_name "sha512.ke")
+set(package_script "./sha512-runner sha512 eyrie-rt loader.bin")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
+
+# eapp
+# Here, instead of compiling from source, we're just copying the pre-compiled binary
+add_custom_target(${eapp_bin}
+  COMMAND cp ${eapp_bin_path} ${CMAKE_BINARY_DIR}/${eapp_bin}
+  COMMENT "Copying precompiled executable to build directory"
+)
+
+# host
+add_executable(${host_bin} ${host_src})
+target_link_libraries(${host_bin} ${KEYSTONE_LIB_HOST} ${KEYSTONE_LIB_EDGE})
+
+# eyrie
+set(eyrie_files_to_copy .options_log eyrie-rt loader.bin)
+add_eyrie_runtime(${eapp_bin}-eyrie
+  ${eyrie_plugins}
+  ${eyrie_files_to_copy})
+
+# add target for packaging (see keystone.cmake)
+add_keystone_package(${eapp_bin}-package
+  ${package_name}
+  ${package_script}
+  ${eyrie_files_to_copy} ${eapp_bin} ${host_bin})
+
+add_dependencies(${eapp_bin}-package ${eapp_bin}-eyrie)
+
+# add package to the top-level target
+add_dependencies(examples ${eapp_bin}-package)
diff --git a/examples/sha512/eapp/.placeholder b/examples/sha512/eapp/.placeholder
new file mode 100644
index 0000000..e69de29
diff --git a/examples/sha512/host/host.cpp b/examples/sha512/host/host.cpp
new file mode 100644
index 0000000..578c9e8
--- /dev/null
+++ b/examples/sha512/host/host.cpp
@@ -0,0 +1,67 @@
+//******************************************************************************
+// Copyright (c) 2018, The Regents of the University of California (Regents).
+// All Rights Reserved. See LICENSE for license details.
+//------------------------------------------------------------------------------
+#include <getopt.h>
+#include <cstdio>
+#include <iostream>
+#include "edge/edge_call.h"
+#include "host/keystone.h"
+
+using namespace Keystone;
+
+int main(int argc, char** argv) {
+  if (argc < 4 || argc > 8) {
+    printf(
+        "Usage: %s <eapp> <runtime> <loader.bin> [--utm-size SIZE(K)] [--freemem-size "
+        "SIZE(K)]\n",
+        argv[0]);
+    return 0;
+  }
+
+  char* eapp_file = argv[1];
+  char* rt_file   = argv[2];
+  char* ld_file   = argv[3];
+
+  size_t untrusted_size = 2 * 1024 * 1024;
+  size_t freemem_size   = 48 * 1024 * 1024;
+
+  static struct option long_options[] = {
+      {"utm-size", required_argument, 0, 'u'},
+      {"freemem-size", required_argument, 0, 'f'},
+      {0, 0, 0, 0}};
+
+
+  int c;
+  int opt_index = 3;
+  while (1) {
+    c = getopt_long(argc, argv, "u:f:", long_options, &opt_index);
+    if (c == -1) break;
+
+    switch (c) {
+      case 0:
+        break;
+      case 'u':
+        untrusted_size = atoi(optarg) * 1024;
+        break;
+      case 'f':
+        freemem_size = atoi(optarg) * 1024;
+        break;
+    }
+  }
+
+  Keystone::Enclave enclave;
+  Keystone::Params params;
+
+  params.setFreeMemSize(freemem_size);
+  params.setUntrustedSize(untrusted_size);
+  enclave.init(eapp_file, rt_file, ld_file, params);
+  enclave.registerOcallDispatch(incoming_call_dispatch);
+  edge_call_init_internals(
+      (uintptr_t)enclave.getSharedBuffer(), enclave.getSharedBufferSize());
+
+  uintptr_t encl_ret;
+  enclave.run(&encl_ret);
+  return 0;
+}
+
diff --git a/examples/tests/CMakeLists.txt b/examples/tests/CMakeLists.txt
index 2112642..e1c5b76 100644
--- a/examples/tests/CMakeLists.txt
+++ b/examples/tests/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(host_bin test-runner)
 set(host_src test-runner.cpp edge_wrapper.cpp)
-set(eyrie_plugins "none")
+set(eyrie_plugins "io_syscall linux_syscall env_setup")
 
 set(package_name "tests.ke")
 set(package_script "./run-test.sh")
@@ -15,7 +15,6 @@ set(all_test_bins
   test-malloc
   test-long-nop
   test-fibonacci
-  test-fib-bench
   test-attestation
   test-untrusted
   test-data-sealing)
diff --git a/overlays/keystone/configs/linux64-defconfig b/overlays/keystone/configs/linux64-defconfig
index f6e0086..1e7b853 100644
--- a/overlays/keystone/configs/linux64-defconfig
+++ b/overlays/keystone/configs/linux64-defconfig
@@ -125,3 +125,4 @@ CONFIG_DEBUG_PLIST=y
 CONFIG_DEBUG_SG=y
 # CONFIG_RCU_TRACE is not set
 CONFIG_RCU_EQS_DEBUG=y
+CONFIG_CMDLINE="root=/dev/vda rw init=/bin/bash"
diff --git a/overlays/keystone/configs/riscv64_generic_defconfig b/overlays/keystone/configs/riscv64_generic_defconfig
index 89e02f3..d244574 100644
--- a/overlays/keystone/configs/riscv64_generic_defconfig
+++ b/overlays/keystone/configs/riscv64_generic_defconfig
@@ -9,7 +9,7 @@ BR2_CCACHE_INITIAL_SETUP="-M0 -F0"
 BR2_GLOBAL_PATCH_DIR="$(BR2_EXTERNAL_KEYSTONE_PATH)/patches"
 BR2_PER_PACKAGE_DIRECTORIES=y
 BR2_SSP_NONE=y
-BR2_TARGET_GENERIC_ROOT_PASSWD="sifive"
+BR2_TARGET_GENERIC_ROOT_PASSWD=""
 BR2_SYSTEM_BIN_SH_BASH=y
 BR2_SYSTEM_DHCP="eth0"
 BR2_ROOTFS_OVERLAY="/invalid"
